src/graphnet/data/extractors/icecube/i3truthextractor.py

```diff
@@ -78,15 +84,85 @@ def __init__(
             self._borders = [border_xy, border_z]
         else:
             self._borders = borders
+
+        self._extend_boundary = extend_boundary
         self._mctree = mctree
 
+    def set_gcd(self, i3_file: str, gcd_file: Optional[str] = None) -> None:
```

Comment:
Limiting the adjustments of`set_gcd` to only `I3TruthExtractor` is a good idea, as the modifications are only needed here. But instead of copying the original function, let's solve this through inheritance. Here is some pseudo code that I think will be identical in function but omit the duplication:

```python
def set_gcd(self, i3_file: str, gcd_file: Optional[str] = None) -> None:
    """Extract GFrame and CFrame from i3/gcd-file pair.
           Information from these frames will be set as member variables of
           `I3Extractor.`
        Args:
            i3_file: Path to i3 file that is being converted.
            gcd_file: Path to GCD file. Defaults to None. If no GCD file is
                      given, the method will attempt to find C and G frames in
                      the i3 file instead. If either one of those are not
                      present, `RuntimeErrors` will be raised.
    """
    # Parent implementation of `set_gcd`
    super().set_gcd(i3_file = i3_file,
                    gcd_file = gcd_file)
    
    # Modifications specific to I3TruthExtractor
    # These modifications are needed to identify starting events
    coordinates = []
    for _, g in self._gcd_dict.items():
        if g.position.z > 1200:
            continue  # We want to exclude icetop
        coordinates.append([g.position.x, g.position.y, g.position.z])
    coordinates = np.array(coordinates)

    if self._extend_boundary != 0.0:
        center = np.mean(coordinates, axis=0)
        d = coordinates - center
        norms = np.linalg.norm(d, axis=1, keepdims=True)
        dn = d / norms
        coordinates = coordinates + dn * self._extend_boundary

    hull = ConvexHull(coordinates)

    self.hull = hull
    self.delaunay = Delaunay(coordinates[self.hull.vertices])
```

Reply:
Good point! Made these changes in the last commit.